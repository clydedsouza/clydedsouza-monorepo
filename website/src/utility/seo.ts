import type { Metadata } from "next";

export interface ISeo {
  title?: string;
  description?: string;
  image?: string;
}

export const WEBSITE_BASE_URL = "https://clydedsouza.net/";

export const generateMetadata = (props: ISeo): Metadata => {
  const titlePrefix = props.title ? `${props.title} — ` : "";
  const title = `${titlePrefix}Clyde D'Souza — Software Engineer and Author`;

  const defaultDescription =
    "Clyde D'Souza is software engineer and author based in Auckland, New Zealand. Feel free to send him a tweet @ClydeDz";
  const description = props.description ?? defaultDescription;

  const defaultOgImage = "https://files.clydedsouza.net/site/ogimage.png";
  const image = props.image ?? defaultOgImage;

  const metadata: Metadata = {
    title,
    description,
    applicationName: "clydedsouza.net",
    openGraph: {
      title,
      description,
      type: "website",
      locale: "en_US",
      images: [{ url: image }],
    },
    twitter: {
      title,
      description,
      creator: "@clydedz",
      images: [{ url: image }],
    },
    authors: {
      name: "Clyde D'Souza",
      url: WEBSITE_BASE_URL,
    },
    keywords:
      "Clyde D'Souza, @ClydeDz, full-stack software engineer, senior software engineer, passionate front end developer, passionate back end developer, full stack developer, author, childrens bedtime stories, eye for ui, ux, Mumbai, India, IND, Auckland, New Zealand, NZ",
    themeColor: "#009688",
    manifest: `${WEBSITE_BASE_URL}manifest.json`,
    icons: [
      {
        rel: "icon",
        url: `${WEBSITE_BASE_URL}favicon.ico`,
      },
      {
        rel: "icon",
        url: "https://files.clydedsouza.net/device-icons/apple-touch-icon-180x180.png",
        sizes: "192x192",
      },
      {
        rel: "apple-touch-icon",
        url: "https://files.clydedsouza.net/device-icons/apple-touch-icon-57x57.png",
        sizes: "57x57",
      },
      {
        rel: "apple-touch-icon",
        url: "https://files.clydedsouza.net/device-icons/apple-touch-icon-114x114.png",
        sizes: "114x114",
      },
      {
        rel: "apple-touch-icon",
        url: "https://files.clydedsouza.net/device-icons/apple-touch-icon-72x72.png",
        sizes: "72x72",
      },
      {
        rel: "apple-touch-icon",
        url: "https://files.clydedsouza.net/device-icons/apple-touch-icon-60x60.png",
        sizes: "60x60",
      },
      {
        rel: "apple-touch-icon",
        url: "https://files.clydedsouza.net/device-icons/apple-touch-icon-120x120.png",
        sizes: "120x120",
      },
      {
        rel: "apple-touch-icon",
        url: "https://files.clydedsouza.net/device-icons/apple-touch-icon-76x76.png",
        sizes: "76x76",
      },
      {
        rel: "apple-touch-icon",
        url: "https://files.clydedsouza.net/device-icons/apple-touch-icon-152x152.png",
        sizes: "152x152",
      },
      {
        rel: "apple-touch-icon",
        url: "https://files.clydedsouza.net/device-icons/apple-touch-icon-180x180.png",
        sizes: "180x180",
      },
    ],
  };
  return metadata;
};
