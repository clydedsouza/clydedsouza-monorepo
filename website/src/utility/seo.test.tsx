import { render } from "@testing-library/react";
import { generateMetadata } from "./seo";

describe("SEO", () => {
  it("should render default seo information", () => {
    const meta = generateMetadata({
      title: undefined,
      description: undefined,
      image: undefined,
    });
    const { container } = render(<>{JSON.stringify(meta)}</>);
    expect(container).toMatchSnapshot();
  });

  it("should render seo information when test page title is passed", () => {
    const meta = generateMetadata({
      title: "Test page",
      description: undefined,
      image: undefined,
    });
    const { container } = render(<>{JSON.stringify(meta)}</>);
    expect(container).toMatchSnapshot();
  });

  it("should render seo information when test description is passed", () => {
    const meta = generateMetadata({
      title: undefined,
      description: "Test description",
      image: undefined,
    });
    const { container } = render(<>{JSON.stringify(meta)}</>);
    expect(container).toMatchSnapshot();
  });

  it("should render seo information when test image is passed", () => {
    const meta = generateMetadata({
      title: undefined,
      description: undefined,
      image: "test.jpg",
    });
    const { container } = render(<>{JSON.stringify(meta)}</>);
    expect(container).toMatchSnapshot();
  });
});
