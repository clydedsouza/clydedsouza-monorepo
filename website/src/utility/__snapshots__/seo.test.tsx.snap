// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SEO should render default seo information 1`] = `
<div>
  {"title":"Clyde D'Souza — Software Engineer and Author","description":"Clyde D'Souza is software engineer and author based in Auckland, New Zealand. Feel free to send him a tweet @ClydeDz","applicationName":"clydedsouza.net","openGraph":{"title":"Clyde D'Souza — Software Engineer and Author","description":"Clyde D'Souza is software engineer and author based in Auckland, New Zealand. Feel free to send him a tweet @ClydeDz","type":"website","locale":"en_US","images":[{"url":"https://files.clydedsouza.net/site/ogimage.png"}]},"twitter":{"title":"Clyde D'Souza — Software Engineer and Author","description":"Clyde D'Souza is software engineer and author based in Auckland, New Zealand. Feel free to send him a tweet @ClydeDz","creator":"@clydedz","images":[{"url":"https://files.clydedsouza.net/site/ogimage.png"}]},"authors":{"name":"Clyde D'Souza","url":"https://clydedsouza.net/"},"keywords":"Clyde D'Souza, @ClydeDz, full-stack software engineer, senior software engineer, passionate front end developer, passionate back end developer, full stack developer, author, childrens bedtime stories, eye for ui, ux, Mumbai, India, IND, Auckland, New Zealand, NZ","themeColor":"#009688","manifest":"https://clydedsouza.net/manifest.json","verification":{"other":{"msvalidate.01":"8ABC4E1B7DBE4391CF265948730B9299","msapplication-TileColor":"#009688","msapplication-TileImage":"https://files.clydedsouza.net/device-icons/apple-touch-icon-180x180.png"}},"icons":[{"rel":"icon","url":"https://clydedsouza.net/favicon.ico"},{"rel":"icon","url":"https://files.clydedsouza.net/device-icons/apple-touch-icon-180x180.png","sizes":"192x192"},{"rel":"apple-touch-icon","url":"https://files.clydedsouza.net/device-icons/apple-touch-icon-57x57.png","sizes":"57x57"},{"rel":"apple-touch-icon","url":"https://files.clydedsouza.net/device-icons/apple-touch-icon-114x114.png","sizes":"114x114"},{"rel":"apple-touch-icon","url":"https://files.clydedsouza.net/device-icons/apple-touch-icon-72x72.png","sizes":"72x72"},{"rel":"apple-touch-icon","url":"https://files.clydedsouza.net/device-icons/apple-touch-icon-60x60.png","sizes":"60x60"},{"rel":"apple-touch-icon","url":"https://files.clydedsouza.net/device-icons/apple-touch-icon-120x120.png","sizes":"120x120"},{"rel":"apple-touch-icon","url":"https://files.clydedsouza.net/device-icons/apple-touch-icon-76x76.png","sizes":"76x76"},{"rel":"apple-touch-icon","url":"https://files.clydedsouza.net/device-icons/apple-touch-icon-152x152.png","sizes":"152x152"},{"rel":"apple-touch-icon","url":"https://files.clydedsouza.net/device-icons/apple-touch-icon-180x180.png","sizes":"180x180"}]}
</div>
`;

exports[`SEO should render seo information when test description is passed 1`] = `
<div>
  {"title":"Clyde D'Souza — Software Engineer and Author","description":"Test description","applicationName":"clydedsouza.net","openGraph":{"title":"Clyde D'Souza — Software Engineer and Author","description":"Test description","type":"website","locale":"en_US","images":[{"url":"https://files.clydedsouza.net/site/ogimage.png"}]},"twitter":{"title":"Clyde D'Souza — Software Engineer and Author","description":"Test description","creator":"@clydedz","images":[{"url":"https://files.clydedsouza.net/site/ogimage.png"}]},"authors":{"name":"Clyde D'Souza","url":"https://clydedsouza.net/"},"keywords":"Clyde D'Souza, @ClydeDz, full-stack software engineer, senior software engineer, passionate front end developer, passionate back end developer, full stack developer, author, childrens bedtime stories, eye for ui, ux, Mumbai, India, IND, Auckland, New Zealand, NZ","themeColor":"#009688","manifest":"https://clydedsouza.net/manifest.json","verification":{"other":{"msvalidate.01":"8ABC4E1B7DBE4391CF265948730B9299","msapplication-TileColor":"#009688","msapplication-TileImage":"https://files.clydedsouza.net/device-icons/apple-touch-icon-180x180.png"}},"icons":[{"rel":"icon","url":"https://clydedsouza.net/favicon.ico"},{"rel":"icon","url":"https://files.clydedsouza.net/device-icons/apple-touch-icon-180x180.png","sizes":"192x192"},{"rel":"apple-touch-icon","url":"https://files.clydedsouza.net/device-icons/apple-touch-icon-57x57.png","sizes":"57x57"},{"rel":"apple-touch-icon","url":"https://files.clydedsouza.net/device-icons/apple-touch-icon-114x114.png","sizes":"114x114"},{"rel":"apple-touch-icon","url":"https://files.clydedsouza.net/device-icons/apple-touch-icon-72x72.png","sizes":"72x72"},{"rel":"apple-touch-icon","url":"https://files.clydedsouza.net/device-icons/apple-touch-icon-60x60.png","sizes":"60x60"},{"rel":"apple-touch-icon","url":"https://files.clydedsouza.net/device-icons/apple-touch-icon-120x120.png","sizes":"120x120"},{"rel":"apple-touch-icon","url":"https://files.clydedsouza.net/device-icons/apple-touch-icon-76x76.png","sizes":"76x76"},{"rel":"apple-touch-icon","url":"https://files.clydedsouza.net/device-icons/apple-touch-icon-152x152.png","sizes":"152x152"},{"rel":"apple-touch-icon","url":"https://files.clydedsouza.net/device-icons/apple-touch-icon-180x180.png","sizes":"180x180"}]}
</div>
`;

exports[`SEO should render seo information when test image is passed 1`] = `
<div>
  {"title":"Clyde D'Souza — Software Engineer and Author","description":"Clyde D'Souza is software engineer and author based in Auckland, New Zealand. Feel free to send him a tweet @ClydeDz","applicationName":"clydedsouza.net","openGraph":{"title":"Clyde D'Souza — Software Engineer and Author","description":"Clyde D'Souza is software engineer and author based in Auckland, New Zealand. Feel free to send him a tweet @ClydeDz","type":"website","locale":"en_US","images":[{"url":"test.jpg"}]},"twitter":{"title":"Clyde D'Souza — Software Engineer and Author","description":"Clyde D'Souza is software engineer and author based in Auckland, New Zealand. Feel free to send him a tweet @ClydeDz","creator":"@clydedz","images":[{"url":"test.jpg"}]},"authors":{"name":"Clyde D'Souza","url":"https://clydedsouza.net/"},"keywords":"Clyde D'Souza, @ClydeDz, full-stack software engineer, senior software engineer, passionate front end developer, passionate back end developer, full stack developer, author, childrens bedtime stories, eye for ui, ux, Mumbai, India, IND, Auckland, New Zealand, NZ","themeColor":"#009688","manifest":"https://clydedsouza.net/manifest.json","verification":{"other":{"msvalidate.01":"8ABC4E1B7DBE4391CF265948730B9299","msapplication-TileColor":"#009688","msapplication-TileImage":"https://files.clydedsouza.net/device-icons/apple-touch-icon-180x180.png"}},"icons":[{"rel":"icon","url":"https://clydedsouza.net/favicon.ico"},{"rel":"icon","url":"https://files.clydedsouza.net/device-icons/apple-touch-icon-180x180.png","sizes":"192x192"},{"rel":"apple-touch-icon","url":"https://files.clydedsouza.net/device-icons/apple-touch-icon-57x57.png","sizes":"57x57"},{"rel":"apple-touch-icon","url":"https://files.clydedsouza.net/device-icons/apple-touch-icon-114x114.png","sizes":"114x114"},{"rel":"apple-touch-icon","url":"https://files.clydedsouza.net/device-icons/apple-touch-icon-72x72.png","sizes":"72x72"},{"rel":"apple-touch-icon","url":"https://files.clydedsouza.net/device-icons/apple-touch-icon-60x60.png","sizes":"60x60"},{"rel":"apple-touch-icon","url":"https://files.clydedsouza.net/device-icons/apple-touch-icon-120x120.png","sizes":"120x120"},{"rel":"apple-touch-icon","url":"https://files.clydedsouza.net/device-icons/apple-touch-icon-76x76.png","sizes":"76x76"},{"rel":"apple-touch-icon","url":"https://files.clydedsouza.net/device-icons/apple-touch-icon-152x152.png","sizes":"152x152"},{"rel":"apple-touch-icon","url":"https://files.clydedsouza.net/device-icons/apple-touch-icon-180x180.png","sizes":"180x180"}]}
</div>
`;

exports[`SEO should render seo information when test page title is passed 1`] = `
<div>
  {"title":"Test page — Clyde D'Souza — Software Engineer and Author","description":"Clyde D'Souza is software engineer and author based in Auckland, New Zealand. Feel free to send him a tweet @ClydeDz","applicationName":"clydedsouza.net","openGraph":{"title":"Test page — Clyde D'Souza — Software Engineer and Author","description":"Clyde D'Souza is software engineer and author based in Auckland, New Zealand. Feel free to send him a tweet @ClydeDz","type":"website","locale":"en_US","images":[{"url":"https://files.clydedsouza.net/site/ogimage.png"}]},"twitter":{"title":"Test page — Clyde D'Souza — Software Engineer and Author","description":"Clyde D'Souza is software engineer and author based in Auckland, New Zealand. Feel free to send him a tweet @ClydeDz","creator":"@clydedz","images":[{"url":"https://files.clydedsouza.net/site/ogimage.png"}]},"authors":{"name":"Clyde D'Souza","url":"https://clydedsouza.net/"},"keywords":"Clyde D'Souza, @ClydeDz, full-stack software engineer, senior software engineer, passionate front end developer, passionate back end developer, full stack developer, author, childrens bedtime stories, eye for ui, ux, Mumbai, India, IND, Auckland, New Zealand, NZ","themeColor":"#009688","manifest":"https://clydedsouza.net/manifest.json","verification":{"other":{"msvalidate.01":"8ABC4E1B7DBE4391CF265948730B9299","msapplication-TileColor":"#009688","msapplication-TileImage":"https://files.clydedsouza.net/device-icons/apple-touch-icon-180x180.png"}},"icons":[{"rel":"icon","url":"https://clydedsouza.net/favicon.ico"},{"rel":"icon","url":"https://files.clydedsouza.net/device-icons/apple-touch-icon-180x180.png","sizes":"192x192"},{"rel":"apple-touch-icon","url":"https://files.clydedsouza.net/device-icons/apple-touch-icon-57x57.png","sizes":"57x57"},{"rel":"apple-touch-icon","url":"https://files.clydedsouza.net/device-icons/apple-touch-icon-114x114.png","sizes":"114x114"},{"rel":"apple-touch-icon","url":"https://files.clydedsouza.net/device-icons/apple-touch-icon-72x72.png","sizes":"72x72"},{"rel":"apple-touch-icon","url":"https://files.clydedsouza.net/device-icons/apple-touch-icon-60x60.png","sizes":"60x60"},{"rel":"apple-touch-icon","url":"https://files.clydedsouza.net/device-icons/apple-touch-icon-120x120.png","sizes":"120x120"},{"rel":"apple-touch-icon","url":"https://files.clydedsouza.net/device-icons/apple-touch-icon-76x76.png","sizes":"76x76"},{"rel":"apple-touch-icon","url":"https://files.clydedsouza.net/device-icons/apple-touch-icon-152x152.png","sizes":"152x152"},{"rel":"apple-touch-icon","url":"https://files.clydedsouza.net/device-icons/apple-touch-icon-180x180.png","sizes":"180x180"}]}
</div>
`;
